File: design-pattern-java/app/src/test/java/design/pattern/java/patterns/behavioral/iterator/IteratorTest.java
package design.pattern.java.patterns.behavioral.iterator;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.junit.jupiter.api.Assertions.fail;

import org.junit.jupiter.api.Test;

public class IteratorTest {
  @Test
  public void shouldProduceFibonacciSequenceWhenNextIsCalled() {
    Producer<Integer> fbProducer = new FibonacciProducer();
    Integer current = 1;
    Integer prev = 0;
    while (fbProducer.hasNext()) {
      Integer obtained = fbProducer.next();
      assertEquals(current + prev, obtained);
      current += prev;
      prev = current - prev;
    }
  }

  @Test
  public void shouldThrowExceptionWhenIndexOutOfBound() {
    Producer<Integer> fbProducer = new FibonacciProducer();
    while (fbProducer.hasNext()) {
      fbProducer.next();
    }
    try {
      fbProducer.next();
      fail("Must throw index out of bound exception");
    } catch (IndexOutOfBoundsException e) {
      assertTrue(true);
    } catch (Exception e) {
      fail("Must throw index out of bound exception");
    }
  }
}

File: design-pattern-jvm/app/src/test/java/design/pattern/kotlin/patterns/behavioral/iterator/IteratorTest.kt
package design.pattern.kotlin.patterns.behavioral.iterator

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import java.lang.Exception
import java.lang.IndexOutOfBoundsException

class IteratorTest {
    @Test
    fun shouldProduceFibonacciSequenceWhenNextIsCalled() {
        val fbProducer: Producer<Int> = FibonacciProducer()
        var current = 1
        var prev = 0
        while (fbProducer.hasNext()) {
            val obtained = fbProducer.next()
            Assertions.assertEquals(current + prev, obtained)
            current += prev
            prev = current - prev
        }
    }

    @Test
    fun shouldThrowExceptionWhenIndexOutOfBound() {
        val fbProducer: Producer<Int> = FibonacciProducer()
        while (fbProducer.hasNext()) {
            fbProducer.next()
        }
        try {
            fbProducer.next()
            Assertions.fail<Any>("Must throw index out of bound exception")
        } catch (e: IndexOutOfBoundsException) {
            Assertions.assertTrue(true)
        } catch (e: Exception) {
            Assertions.fail<Any>("Must throw index out of bound exception")
        }
    }
}